Handle start/end positions equal to the root's text content length

robertknight opened the pull request:
Re-implement the `toRange` function to better handle cases where:

- The start or end positions include the end of the root element's text
- The start or end positions are outside the range `[0, root.textContent.length]`
- The root node is empty or consists of just an empty text node

This fixes an issue encountered in Hypothesis' PDF annotation where
creating an annotation would fail if the selection included the last
character of a PDF page's text layer ([original bug report](https://github.com/hypothesis/client/issues/1329)).

This implementation also makes it more obvious which of the two possible
range configurations is used when the `start` or `end` offset occurs at
a text node boundary. In that case, the range could use either the end
of one text node or the start of the next as the boundary point. The
boundary point is always placed at the end of the earlier node in that case.

robertknight committed:
Handle start/end positions equal to the root's text content length

Re-implement the `toRange` function to better handle cases where the
start or end positions include the end of the root element's text, or
the start or end positions are outside the range `[0, root.textContent.length]`.

This fixes an issue encountered in Hypothesis' PDF annotation where
creating an annotation would fail if the selection included the last
character of a PDF page's text layer.

This implementation also makes it more obvious which of the two possible
range configurations is used when the `start` or `end` offset occurs at
a text node boundary. In that case, the range could use either the end
of one text node or the start of the next as the boundary point. The
boundary point is always placed at the end of the earlier node in that case.

robertknight committed:
Add tests for empty root nodes

Add tests to clarify the expected behavior when the root node is empty
or consists only of an empty text node.

robertknight committed:
Add a test for positions that cover the initial text of the root element

robertknight committed:
Stop node iteration when both ends of the range have been found

This is an optimization to avoid unnecessary iteration once a selector
has been resolved.

tilgovi commented on test/index_spec.js:
```diff
@@ -218,5 +218,65 @@ describe('textPosition', () => {
+    it('handles an empty root element', () => {
+      let root = document.createElement('div');
+      assert.throws(() => {
+        toRange(root, {start:0, end: 0})
```
Seems like it shouldn't throw it this case, since `[0, 0)` is a valid, empty range.

tilgovi replied:
I might be convinced it should, but I've gotta think about it a little bit.

robertknight replied:
My thinking here is that there is an implicit invariant in all other cases that the returned range starts and ends in a text node. If there are no text nodes, that's not possible.

The other motivation was simply convenience as it avoids needing to special-case this in the implementation.

tilgovi commented on test/index_spec.js:
```diff
@@ -254,5 +254,19 @@ describe('textPosition', () => {
+
+    it('handles a root element with an empty text node', () => {
+      let root = document.createElement('div');
+      root.appendChild(document.createTextNode())
```
`document.createTextNode()` needs an empty string argument. At least under Chromium, locally, it seems to throw without it.

tilgovi commented:
These tests are great, thank you.

I just pushed a branch that passes all your tests, preserving more of the original structure: https://github.com/tilgovi/dom-anchor-text-position/compare/fix-overruns-and-empty-root?expand=1

It highlights some issues with the ergonomics of `seek()` that we could, perhaps, resolve.

I don't know if I'm just experiencing a personal attachment or if there are real benefits to maintain the composition of building this on top of `seek()`. I'm tempted, maybe, because it provides a place to resolve #6 / tilgovi/dom-seek#1.

Any objection to merging your test changes with my implementation changes, or do you think I'm being sentimental by keeping `seek()`?

robertknight commented:
> Any objection to merging your test changes with my implementation changes, or do you think I'm being sentimental by keeping seek()?

If the updated dom-seek based implementation passes the tests, that at least solves our problem. So if you could merge either of the two approaches, that would be great. One technical issue I noticed in the code is that it assigns to `iter.referenceNode`, which is a read-only property. That's a no-op in non-strict code and an error in strict-code. I'm not certain why it doesn't error in this context - either the code is not being executed in strict mode, or PhantomJS is incorrectly not throwing, or the `referenceNode` property refers to something from the polyfill?

The reason I opted for a different approach originally is that I found it difficult to reason about the correctness when extending the existing implementation. What the code is ultimately trying to do seems fairly simple: Iterate over a sequence of text nodes and determine the node and character offsets, if any, which correspond to two offsets within the concatenated text of all the nodes. However with the previous implementation I found myself needing to read the `NodeIterator` DOM specs to grok how `pointerBeforeReference` works when iterating forwards and backwards, reading the dom-seek implementation to clarify how it works and what happens in various error scenarios (what if there are no text nodes? What if the text length is less than the `where` value? What are the invariants about the state of the iterator afterwards? etc.). So ultimately I reworked it to hopefully make it more "obviously correct" for a future reader.

tilgovi commented:
> One technical issue I noticed in the code is that it assigns to iter.referenceNode, which is a read-only property.

That was a mistake of refactoring. I had a separate variable, like you have, `startContainer`, that was getting assigned. It's not needed and I've fixed that.

> The reason I opted for a different approach originally is that I found it difficult to reason about the correctness when extending the existing implementation.

That's exactly right. The ergonomics are bad and I'm not sure whether they can be salvaged or not. It might be doable, though! Maybe a) throwing if there is no text at all, and b) always leaving the pointer before the reference node, c) fixing the documentation to be clear about overruns.

tilgovi commented:
Just to see what that would look like:

``` javascript
export function toRange(root, selector = {}) {
  if (root === undefined) {
    throw new Error('missing required parameter "root"')
  }

  let document = root.ownerDocument
  let range = document.createRange()
  let iter = createNodeIterator(root, SHOW_TEXT)

  let start = selector.start || 0
  let end = selector.end || start
  let count = seek(iter, start)
  let remainder = start - count

  range.setStart(iter.referenceNode, remainder)

  let length = (end - start) + remainder
  count = seek(iter, length)
  remainder = length - count

  range.setEnd(iter.referenceNode, remainder)

  return range
}
```

I find that really attractive.

tilgovi commented:
Alternatively:

``` javascript
export function toRange(root, selector = {}) {
  if (root === undefined) {
    throw new Error('missing required parameter "root"')
  }

  const document = root.ownerDocument
  const range = document.createRange()
  const iter = createNodeIterator(root, SHOW_TEXT)

  const start = selector.start || 0
  const end = selector.end || start

  const startOffset = start - seek(iter, start);
  const startContainer = iter.referenceNode;

  const remainder = end - start + startOffset;

  const endOffset = remainder - seek(iter, remainder);
  const endContainer = iter.referenceNode;

  range.setStart(startContainer, startOffset)
  range.setEnd(endContainer, endOffset)

  return range
}
```

robertknight committed:
Add missing mandatory argument to `createTextNode`

PhantomJS does not throw if this argument is missing, but modern
browsers do.

This fixes the tests in Chrome.

robertknight committed:
Add a test for what happens when `end` is less than `start`

robertknight committed:
Add note about why `toRange` throws if there are no text nodes

robertknight commented:
I had a go at using the code in the previous comment as a basis. Using the tests in this branch I encountered a few issues:

- If the `start` position refers to a node boundary (see `"toRange returns a range selecting a whole text node"` test), then the first call to `seek` leaves the iterator pointing at the previous node to the one containing the content and so the returned range selects the wrong content. The code in the `master` branch knows about this detail of `dom-seek`'s implementation and handles it explicitly
- If the `end` position refers to the end of the root's content , then the `endOffset` ends up being incorrect and the returned range is empty (see `"toRange returns a range selecting the last text of the root element"` test)
- The cases where the start and end offsets are out of range is not implemented in this sketch explicitly. Those tests currently fail with a generic DOM exception due to an invalid character offset being passed to `setStart` and `setEnd`. Adding checks for this is complicated by the fact that `seek` may leave `iter.referenceNode` pointing an either an element or a text node depending on whether the root has text nodes and whether the passed `whence` value is 0 or not

I'm sure it is possible to resolve these issues, but it does seem to me simpler to avoid the dom-seek abstraction for this particular use case.

I pushed a couple more commits to add the missing argument to `createTextNode` and also add a test for when the `end` offset is less than the `start` offset.

tilgovi commented:
Very sorry if I caused any confusion. My code snippets were aspirational, with the idea that I would fix dom-seek to provide those guarantees.

I would like to unblock you, so I will take a look at what you've done and probably merge something later today. Thanks for all the work on this.

I am somewhat attached to the seek abstraction, generally, because I do observe that iterating over text nodes and counting characters is something we do in a lot of places in Hypothesis / Annotator / dom-anchor-*.

robertknight commented:
> My code snippets were aspirational, with the idea that I would fix dom-seek to provide those guarantees.

Ah, right. I had misunderstood.

> I would like to unblock you, so I will take a look at what you've done and probably merge something later today. Thanks for all the work on this.

Thanks in advance :)

tilgovi commented:
@robertknight it took me longer to get back to the list last week than I thought. Over the weekend, I did have time to experiment a bit.

I've kept all of your commits and added one of my own, implementing `toRange()` as described in my comment, above. That branch is at #8. It is currently pointing at a branch on `dom-seek` that includes proposed changes in tilgovi/dom-seek#6 along with the built sources for testing.

Would this be okay for you?

longhotsummer commented:
@tilgovi Any thoughts on merging this?

tilgovi commented:
@longhotsummer thanks for the nudge. Working on it this week.

tilgovi commented:
I've updated #8, which currently points at an unreleased version of dom-seek. I put a bunch of work into the master branch of dom-seek and I'm looking for feedback around releasing that as dom-seek v5.

The approach keeps the code here simple and puts more of the logic in dom-seek. I decided that if dom-seek is to be useful it should throw on the error cases and make it easier to work with the iterator. Therefore, it has a few new things going for it:

- Throws `RangeError` for empty text and exhausted traversals
- Does not care what the value of the `pointerBeforeReferenceNode` attribute is at the start
- Always leaves `pointerBeforeReferenceNode` to `true` at the end

With these changes, there's no need for error handling here and it's always the case that you can just use the remainder of the seek when calling `Range.prototype.setStart()` or `Range.prototype.setEnd()`.

Any feedback is greatly appreciated.

longhotsummer commented:
Thanks @tilgovi. I don't have a strong opinion on the internal implementation, although from an aesthetics standpoint the code in #8 is very clean and understandable (although #7 isn't too bad either).

From a library user standpoint, a release that works correctly is the most important thing to me. 

Currently I have cut-and-pasted the code from the hypothesis branch (effectively this PR) because it solves a number of edge cases that were hurting my users and it's simple enough that I have a good understanding of how it works. Being able to rely on a working library is preferable to me than cut-and-paste :)

If #8 requires further iteration, perhaps an option is to release this version (that doesn't rely on a new dom-seek) and then iterate on #8 until it's ready, then release that.

tilgovi commented:
I've just published dom-seek v5.0.0 and updated #8 to use it. That branch keeps all the tests added here but modifies them to expect the new errors thrown by this new dom-seek.

@robertknight thank you so much for your patience with this. I hope you don't feel I've trampled on your work by trying to keep dom-seek alive. I think there's value in having that traversal code exist apart from this. I've checked that the hypothesis/client tests run successfully when I update it for the work in #8. If there's anything I can do to further help with testing there, please reach out.

@longhotsummer thanks for your feedback. I'm going to close this and merge #8 to master. I'll work tomorrow on getting a release out.

robertknight commented:
Hi @tilgovi - Thanks for taking the time to get this resolved. As long as the tests are passing and I can understand the changes I'm not especially attached to which implementation is used.

tilgovi commented:
I published v5.0.0 just now! :tada: 

Note that this version removes the dependency on `dom-node-iterator`, but still depends on a modern implementation.

If you're supporting recent evergreen browsers, you should be fine without it. If you're supporting IE11 or Edge < 17 you'll want it.

I also released v5.0.0 of `dom-node-iterator`. It fixes a bug I found during testing and makes all the `NodeIterator` attributes readonly, per the spec. To automatically polyfill the global document, import `dom-node-iterator/auto`. More over at [that repo](https://github.com/tilgovi/dom-node-iterator).

Please report any bugs when testing and I'll resolve them as quickly as a I can.

tilgovi commented:
If we'd rather not have to pollute the global scope, I could reinstate the hard dependency on dom-node-iterator. I can't decide if it's more appropriate for this library to assume a spec-compliant world or to guard against it without polluting it. Feedback welcome.

longhotsummer commented:
Perhaps assume a standards-compliant world, but let the caller pass in a callable to use if they don’t want to use the browser’s implementation.

So the best case just works with no additional dependency, and the case which needs a workaround is still possible and graceful.

tilgovi commented:
I'll leave this alone for now and will look out for any reports of issues using `dom-node-iterator/auto`. The work I put into the latest dom-node-iterator takes care to try to be a little bit more defensive, ensuring that wrappers it installs for IE11 and Edge < 17 still pass `isinstance NodeIterator` checks and that all the attributes are readonly. So, hopefully this is safe enough for now and I'll look out for reports of issues.
